### ISSUE TITLE
Help with grandle

### ISSUE THREAD

**Issue by Stypox** [vitalidze/chromecast-java-api-v2#108]
If I try to add `runtime 'su.litvak.chromecast:api-v2:0.10.6'` to the app's build.grandle under `dependencies {}` and `mavenCentral()` to the root build.grandle under `allprojects {repositories {}}` I get this error (while syncing with Android Studio) `ERROR: Could not find method runtime() for arguments [su.litvak.chromecast:api-v2:0.10.6] on object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.`

I tried changing `runtime` with `implementation`, since I suspected `runtime` didn't exist, but then I get this error on compilation `More than one file was found with OS independent path 'META-INF/ASL2.0'`.

By searching on the internet I found [this](https://stackoverflow.com/questions/44342455/more-than-one-file-was-found-with-os-independent-path-meta-inf-license) but it didn't help, since it would give me this error on compilation:
```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':app:transformClassesWithDexBuilderForDebug'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:103)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:73)
	at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
	at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
	at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
	at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:256)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:249)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:238)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:663)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:597)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: java.lang.RuntimeException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593)
	at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677)
	at java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:720)
	at com.android.ide.common.internal.WaitableExecutor.waitForTasksWithQuickFail(WaitableExecutor.java:146)
	at com.android.build.gradle.internal.transforms.DesugarIncrementalTransformHelper.getIncrementalData(DesugarIncrementalTransformHelper.java:192)
	at com.android.build.gradle.internal.transforms.DesugarIncrementalTransformHelper.lambda$new$0(DesugarIncrementalTransformHelper.java:84)
	at com.android.builder.desugaring.DesugaringGraphs.updateVariant(DesugaringGraphs.java:81)
	at com.android.build.gradle.internal.transforms.DesugarIncrementalTransformHelper.<init>(DesugarIncrementalTransformHelper.java:83)
	at com.android.build.gradle.internal.transforms.DexArchiveBuilderTransform.transform(DexArchiveBuilderTransform.java:320)
	at com.android.build.gradle.internal.pipeline.TransformTask$2.call(TransformTask.java:239)
	at com.android.build.gradle.internal.pipeline.TransformTask$2.call(TransformTask.java:235)
	at com.android.builder.profile.ThreadRecorder.record(ThreadRecorder.java:102)
	at com.android.build.gradle.internal.pipeline.TransformTask.transform(TransformTask.java:230)
	at sun.reflect.GeneratedMethodAccessor101.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
	at org.gradle.api.internal.project.taskfactory.IncrementalTaskAction.doExecute(IncrementalTaskAction.java:50)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:124)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:113)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:95)
	... 33 more
Caused by: java.lang.RuntimeException
	at org.objectweb.asm.ClassVisitor.visitModule(ClassVisitor.java:148)
	at org.objectweb.asm.ClassReader.readModule(ClassReader.java:731)
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:632)
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:500)
	at com.android.builder.desugaring.DesugaringClassAnalyzer.analyze(DesugaringClassAnalyzer.java:144)
	at com.android.builder.desugaring.DesugaringClassAnalyzer.analyzeJar(DesugaringClassAnalyzer.java:92)
	at com.android.builder.desugaring.DesugaringClassAnalyzer.analyze(DesugaringClassAnalyzer.java:63)
	at com.android.build.gradle.internal.transforms.DesugarIncrementalTransformHelper.lambda$getIncrementalData$5(DesugarIncrementalTransformHelper.java:181)
	at java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1424)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)


```

I have no experience at all with grandle. Sorry

**Comment by vitalidze** [comment]
I am not sure what is the proper way to do it in android project. But it works just fine for me with the following `build.gradle` for a java app (Gradle wrapper version is 3.5):

```gradle
group 'su.litvak'
version '1.0-SNAPSHOT'

apply plugin: 'java'

sourceCompatibility = 1.8

repositories {
    mavenCentral()
}

dependencies {
    compile 'su.litvak.chromecast:api-v2:0.10.6'
}
```

In addition I have the following comments:

1. For Android there is an official SDK for ChromeCast. I am pretty sure that the best way to communicate with such devices is to use official SDK. This library is intended to be used on non-mobile devices.

2. If you still want to use it in your Android app then please provide full Android project so I can try loading it myself.

**Comment by zunjae** [comment]
You're betting off using the official SDK. There's too much pain associated with making this work on Android

**Comment by Stypox** [comment]
I'm using this because I like open source software and privacy,  that is free of Google binaries, and the official SDK would require me

**Comment by Stypox** [comment]
I'm using this because I like open source software and privacy, so Google Services do not fit. To get my full Android Studio project just clone [NewPipe](https://github.com/TeamNewPipe/NewPipe) and add the library to grandle. Probably you need to add

**Comment by Stypox** [comment]
I'm using this because I like open source software and  privacy, so Google Services do not fit. To get my full Android Studio project just clone [NewPipe](https://github.com/TeamNewPipe/NewPipe) and add the library to grandle.

**Comment by vitalidze** [comment]
@Stypox I just got it working for that project. To achieve that I did the following:

1. Added two dependencies to the `implementation` configuration in the`app/build.gradle` file:

```gradle
    implementation 'su.litvak.chromecast:api-v2:0.10.6'
    implementation 'org.slf4j:slf4j-api:1.7.25'
```

2. Added hint for the packager (not sure about correct naming) to exclude duplicated files during build to the `android` section (found solution [here](https://stackoverflow.com/questions/44342455/more-than-one-file-was-found-with-os-independent-path-meta-inf-license)):

```gradle
    packagingOptions {
        exclude 'META-INF/ASL2.0'
    }
```

So basically there are two issues:

1. The duplicated license file called `ASL2.0` in the jackson library dependency. This is the one you describe in your first message.

2. Not working slfj4-1.8-beta dependency, which is loaded as the latest dependency. This is restricted using some specific 1.7 version of the slf4j.

Hope this will work for you. Please let me know if it does not.

**Comment by vitalidze** [comment]
@Stypox any update on this?

**Comment by Stypox** [comment]
Yes, it works. Thank you for taking the time to help me, I hope not to have bothered you. I'm sorry I did not get to this 10 days ago but for some reason GitHub didn't send me any notification...
Oh and by the way, the only thing I was missing was the `'org.slf4j:slf4j-api:1.7.25'` dependency.
Again, thank you!

**Comment by grolschie** [comment]
Just a comment. I've been using this library on Android for a few months just using the instructions on the README. It works very well. If you shrink/obfuscate your APK with ProGuard, you will need to add an exclusion to the ProGuard rules file.

**Comment by grolschie** [comment]
Just a comment. I've been using this library on Android for a few months using:
    `implementation 'su.litvak.chromecast:api-v2:0.11.0'`
It works very well. If you shrink/obfuscate your APK with ProGuard, you will need to add an exclusion to the ProGuard rules file.
